// @solana-developers/preset-react/src/generators/react-template/react-template-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactTemplateGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const path = require("path");
const utils_1 = require("../../utils");
function reactTemplateGenerator(tree_1, options_1) {
    return tslib_1.__awaiter(this, arguments, void 0, function* (tree, options, templatePath = '') {
        const substitutions = (0, utils_1.reactApplicationSubstitutions)({
            anchor: options.anchor,
            anchorName: options.anchorName,
            anchorProgram: options.anchorProgram,
            licenseAuthor: options.licenseAuthor,
            name: options.webName,
            npmScope: options.npmScope,
            preset: options.preset,
        });
        (0, devkit_1.generateFiles)(tree, path.join(__dirname, 'files', options.template, templatePath), options.directory, Object.assign(Object.assign(Object.assign({}, options), substitutions), { fileNameUnderscore: substitutions.fileName.replace(/-/g, '_') }));
    });
}
exports.reactTemplateGenerator = reactTemplateGenerator;
exports.default = reactTemplateGenerator;
