// @nx/next/src/generators/init/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nextInitGeneratorInternal = exports.nextInitGenerator = void 0;
const devkit_1 = require("@nx/devkit");
const add_plugin_1 = require("@nx/devkit/src/utils/add-plugin");
const versions_1 = require("@nx/react/src/utils/versions");
const add_gitignore_entry_1 = require("../../utils/add-gitignore-entry");
const versions_2 = require("../../utils/versions");
function updateDependencies(host, schema) {
    const tasks = [];
    tasks.push((0, devkit_1.removeDependenciesFromPackageJson)(host, ['@nx/next'], []));
    tasks.push((0, devkit_1.addDependenciesToPackageJson)(host, {
        next: versions_2.nextVersion,
        react: versions_1.reactVersion,
        'react-dom': versions_1.reactDomVersion,
    }, {
        '@nx/next': versions_2.nxVersion,
    }, undefined, schema.keepExistingVersions));
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
function nextInitGenerator(tree, schema) {
    return nextInitGeneratorInternal(tree, { addPlugin: false, ...schema });
}
exports.nextInitGenerator = nextInitGenerator;
async function nextInitGeneratorInternal(host, schema) {
    const nxJson = (0, devkit_1.readNxJson)(host);
    const addPluginDefault = process.env.NX_ADD_PLUGINS !== 'false' &&
        nxJson.useInferencePlugins !== false;
    schema.addPlugin ??= addPluginDefault;
    if (schema.addPlugin) {
        const { createNodes } = await Promise.resolve().then(() => require('../../plugins/plugin'));
        await (0, add_plugin_1.addPlugin)(host, await (0, devkit_1.createProjectGraphAsync)(), '@nx/next/plugin', createNodes, {
            startTargetName: ['start', 'next:start', 'next-start'],
            buildTargetName: ['build', 'next:build', 'next-build'],
            devTargetName: ['dev', 'next:dev', 'next-dev'],
            serveStaticTargetName: [
                'serve-static',
                'next:serve-static',
                'next-serve-static',
            ],
        }, schema.updatePackageScripts);
    }
    (0, add_gitignore_entry_1.addGitIgnoreEntry)(host);
    let installTask = () => { };
    if (!schema.skipPackageJson) {
        installTask = updateDependencies(host, schema);
    }
    return installTask;
}
exports.nextInitGeneratorInternal = nextInitGeneratorInternal;
exports.default = nextInitGenerator;
