// @nx/next/src/generators/application/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const devkit_1 = require("@nx/devkit");
const project_name_and_root_utils_1 = require("@nx/devkit/src/generators/project-name-and-root-utils");
const eslint_1 = require("@nx/eslint");
const assertion_1 = require("@nx/react/src/utils/assertion");
async function normalizeOptions(host, options) {
    const { projectName: appProjectName, projectRoot: appProjectRoot, projectNameAndRootFormat, } = await (0, project_name_and_root_utils_1.determineProjectNameAndRootOptions)(host, {
        name: options.name,
        projectType: 'application',
        directory: options.directory,
        projectNameAndRootFormat: options.projectNameAndRootFormat,
        rootProject: options.rootProject,
        callingGenerator: '@nx/next:application',
    });
    options.rootProject = appProjectRoot === '.';
    options.projectNameAndRootFormat = projectNameAndRootFormat;
    const nxJson = (0, devkit_1.readNxJson)(host);
    const addPlugin = process.env.NX_ADD_PLUGINS !== 'false' &&
        nxJson.useInferencePlugins !== false;
    options.addPlugin ??= addPlugin;
    let e2eWebServerTarget = options.addPlugin ? 'start' : 'serve';
    if (options.addPlugin) {
        if (nxJson.plugins) {
            for (const plugin of nxJson.plugins) {
                if (typeof plugin === 'object' &&
                    plugin.plugin === '@nx/next/plugin' &&
                    plugin.options.startTargetName) {
                    e2eWebServerTarget = plugin.options
                        .startTargetName;
                }
            }
        }
    }
    let e2ePort = options.addPlugin ? 3000 : 4200;
    if (nxJson.targetDefaults?.[e2eWebServerTarget] &&
        (nxJson.targetDefaults?.[e2eWebServerTarget].options?.port ||
            nxJson.targetDefaults?.[e2eWebServerTarget].options?.env?.PORT)) {
        e2ePort =
            nxJson.targetDefaults?.[e2eWebServerTarget].options?.port ||
                nxJson.targetDefaults?.[e2eWebServerTarget].options?.env?.PORT;
    }
    const e2eProjectName = options.rootProject ? 'e2e' : `${appProjectName}-e2e`;
    const e2eProjectRoot = options.rootProject ? 'e2e' : `${appProjectRoot}-e2e`;
    const e2eWebServerAddress = `http://localhost:${e2ePort}`;
    const name = (0, devkit_1.names)(options.name).fileName;
    const outputPath = (0, devkit_1.joinPathFragments)('dist', appProjectRoot, ...(options.rootProject ? [name] : []));
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    const fileName = 'index';
    const appDir = options.appDir ?? true;
    const src = options.src ?? true;
    const styledModule = /^(css|scss|less|tailwind)$/.test(options.style)
        ? null
        : options.style;
    (0, assertion_1.assertValidStyle)(options.style);
    return {
        ...options,
        appDir,
        src,
        appProjectRoot,
        e2eProjectName,
        e2eProjectRoot,
        e2eWebServerAddress,
        e2eWebServerTarget,
        e2ePort,
        e2eTestRunner: options.e2eTestRunner || 'playwright',
        fileName,
        linter: options.linter || eslint_1.Linter.EsLint,
        name,
        outputPath,
        parsedTags,
        projectName: appProjectName,
        style: options.style || 'css',
        styledModule,
        unitTestRunner: options.unitTestRunner || 'jest',
    };
}
exports.normalizeOptions = normalizeOptions;
