// @solana-developers/preset-react/src/utils/generate-react-application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateReactApplication = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const eslint_1 = require("@nx/eslint");
const react_1 = require("@nx/react");
const path_1 = require("path");
function generateReactApplication(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        yield (0, react_1.applicationGenerator)(tree, {
            name: options.webName,
            directory: options.directory,
            style: 'css',
            skipFormat: options.skipFormat,
            projectNameAndRootFormat: 'as-provided',
            unitTestRunner: 'none',
            e2eTestRunner: 'none',
            linter: eslint_1.Linter.EsLint,
            pascalCaseFiles: false,
            classComponent: false,
            routing: true,
            strict: true,
            rootProject: false,
            bundler: 'webpack',
        });
        const project = (0, devkit_1.getProjects)(tree).get(options.webName);
        (0, devkit_1.updateJson)(tree, (0, path_1.join)(project.root, 'project.json'), (json) => {
            json.targets.serve.options.port = options.port;
            return json;
        });
        return (0, devkit_1.getProjects)(tree).get(options.webName);
    });
}
exports.generateReactApplication = generateReactApplication;
