// @solana-developers/preset-react/src/utils/setup-anchor-react-feature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupAnchorReactFeature = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const preset_anchor_1 = require("@solana-developers/preset-anchor");
const react_feature_1 = require("../generators/react-feature");
function setupAnchorReactFeature(tree, options, keypair) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (options.anchor === 'none' || (0, devkit_1.getProjects)(tree).has(options.anchorName)) {
            return;
        }
        const featureName = options.anchor;
        const feature = react_feature_1.features.find((feature) => feature.toString() === `anchor-${featureName}`);
        if (!feature) {
            throw new Error(`Invalid anchor feature: ${featureName}`);
        }
        // Create the Anchor application
        yield (0, preset_anchor_1.anchorApplicationGenerator)(tree, { name: options.anchorName, programName: options.anchorProgram, skipFormat: true }, keypair);
        // Create the React feature
        yield (0, react_feature_1.reactFeatureGenerator)(tree, {
            name: options.anchorProgram,
            anchorName: options.anchorName,
            webName: options.webName,
            skipFormat: true,
            feature,
            ui: options.ui,
        }, keypair);
    });
}
exports.setupAnchorReactFeature = setupAnchorReactFeature;
