// @solana-developers/preset-next/src/generators/next-template/next-template-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nextTemplateGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const path = require("path");
const utils_1 = require("../../utils");
function nextTemplateGenerator(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const substitutions = (0, utils_1.nextApplicationSubstitutions)({
            anchor: options.anchor,
            anchorName: options.anchorName,
            anchorProgram: options.anchorProgram,
            name: options.webName,
            npmScope: options.npmScope,
        });
        (0, devkit_1.generateFiles)(tree, path.join(__dirname, 'files', options.template), options.directory, Object.assign(Object.assign(Object.assign({}, options), substitutions), { fileNameUnderscore: substitutions.fileName.replace(/-/g, '_') }));
    });
}
exports.nextTemplateGenerator = nextTemplateGenerator;
exports.default = nextTemplateGenerator;
