// @solana-developers/preset-common/src/generators/common-template/common-template-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.commonTemplateGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const path = require("path");
const generic_substitutions_1 = require("../../utils/generic-substitutions");
function commonTemplateGenerator(tree, options, pm) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const substitutions = (0, generic_substitutions_1.genericSubstitutions)({
            anchor: options.anchor,
            anchorName: options.anchorName,
            anchorProgram: options.anchorProgram,
            licenseAuthor: options.licenseAuthor,
            name: options.webName,
            npmScope: options.npmScope,
            preset: options.preset,
        }, pm);
        (0, devkit_1.generateFiles)(tree, path.join(__dirname, 'files', options.template), options.directory, Object.assign(Object.assign(Object.assign({}, options), substitutions), { fileNameUnderscore: substitutions.fileName.replace(/-/g, '_') }));
    });
}
exports.commonTemplateGenerator = commonTemplateGenerator;
exports.default = commonTemplateGenerator;
