// @solana-developers/preset-next/src/utils/generate-next-application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateNextApplication = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const eslint_1 = require("@nx/eslint");
const next_1 = require("@nx/next");
const path_1 = require("path");
function generateNextApplication(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        yield (0, next_1.applicationGenerator)(tree, {
            src: false,
            name: options.webName,
            style: 'css',
            skipFormat: true,
            projectNameAndRootFormat: 'as-provided',
            unitTestRunner: 'none',
            e2eTestRunner: 'none',
            linter: eslint_1.Linter.EsLint,
            rootProject: false,
        });
        const project = (0, devkit_1.getProjects)(tree).get(options.webName);
        (0, devkit_1.updateJson)(tree, (0, path_1.join)(project.root, 'project.json'), (json) => {
            json.targets.serve.options.port = options.port;
            return json;
        });
        const packageName = '@/';
        (0, devkit_1.updateJson)(tree, 'tsconfig.base.json', (json) => {
            json.compilerOptions.paths[`${packageName}*`] = [`./${options.webName}/*`];
            return json;
        });
        (0, devkit_1.updateJson)(tree, (0, path_1.join)(project.root, '.eslintrc.json'), (json) => {
            json.overrides = [
                ...json.overrides,
                {
                    files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
                    rules: {
                        '@nx/enforce-module-boundaries': [
                            'error',
                            {
                                allow: [packageName],
                            },
                        ],
                    },
                },
            ];
            return json;
        });
        // generate the root workspace files (specific to the NextJS preset generator)
        (0, devkit_1.generateFiles)(tree, (0, path_1.join)(__dirname, '../generators/next-template/files', 'workspace-root'), '.', Object.assign({}, options));
        return project;
    });
}
exports.generateNextApplication = generateNextApplication;
